### PR TITLE
Add deprecation policy

### PR DESCRIPTION
**Pull request by thockin**
This proposed a written policy for deprecation within Kubernetes.

@bgrant0607 @lavalamp @matchstick @erictune @smarterclayton for review

@devin-donnelly I have no idea where to actually put this or where to link it into the tree.  Thoughts?

@sarahnovotny FYI

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1856)

### CODE REVIEW

File: docs/deprecation_policy.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+that do not exist in some versions.
+
+For example, an object can be written as v1 and then read back as v2 and
+converted to v1, and the resulting v1 resource will be identical to the
+original.  The representation in v2 might be different from v1, but the system
+knows how to convert between them in both directions.
+
+Rule #3: Other than the “current” API version, older API versions must be
+supported after their announced deprecation for a duration of no less than:
+   * GA: 1 year or 2 releases (whichever is longer)
```
Comment (thockin) [inline docs/deprecation_policy.md]: From @lavalamp out-of-band: Given that we have many API groups, and they will be versioning at different rates, it's easy to imagine a situation where we're removing a major version from some group every release. If the purpose of this is to reduce the maintenance burden on users, what about a rule that's something like, "GA versions will only be removed during the latter half of the calendar year" or “GA versions will only be removed in odd-numbered releases”?

File: docs/deprecation_policy.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+behavior get removed.
+
+### Fields of REST resources
+
+As with whole REST resources, an individual field which was present in API v1
+must exist and function until API v1 is removed.  Unlike whole resources, the
+v2 APIs may choose a different representation for the field, as long as it can
+be round-tripped.  For example a v1 field named “magnitude” which was
+deprecated might be named “deprecatedMagnitude” in API v2.  When v1 is
+eventually removed, the deprecated field can be removed from v2.
```
Comment (thockin) [inline docs/deprecation_policy.md]: From @lavalamp out-of-band: Does it need to wait until v3? The text as written would work better in a world where we have fine-grained versions, I think...

File: docs/deprecation_policy.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+| X+5     | v1, v2, v2beta2      | * v2beta1 is removed, “action required” relnote<br> * v2beta2 is deprecated, “action required” relnote<br> * v1 is deprecated, “action required” relnote |
+| X+6     | v1, v2               | * v2beta2 is removed, “action required” relnote |
+| X+7     | v1, v2               | |
+| X+8     | v1, v2               | |
+| X+9     | v2                   | * v1 is removed, “action required” relnote |
+
+### REST resources (aka API objects)
+
+Consider a hypothetical REST resource named Widget, which was present in API v1
+in the above timeline, and which we want to deprecate.  We [document](http://kubernetes.io/docs/deprecated/)
```
Comment (thockin) [inline docs/deprecation_policy.md]: @devin-donnelly We need a place to host deprecation notices, and maybe a template for it.  I can help with the template, if you can tell me where to put them and where to link them from.

File: docs/deprecation_policy.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+   * REST resources (aka API objects)
+   * Fields of REST resources
+   * Enumerated or constant values
+These rules are enforced between official releases, not between
+arbitrary commits to master or release branches.
+
+Rule #1: API elements may only be removed by incrementing the version of the
+API group.
+
+Once an API element has been added to an API group at a particular version, it
```
Comment (smarterclayton) [inline docs/deprecation_policy.md]: Would go further and describe "it must not change in behavior".  Removal

File: docs/deprecation_policy.md (hunk 1 above)
Comment (smarterclayton) [inline docs/deprecation_policy.md]: Hrm.  We may not have to solve this before we actually hit the problem.  Could be vaguer.

**Comment by bgrant0607** [comment]
@thockin This is intended to be read by end users or contributors?

**Comment by sarahnovotny** [comment]
@bgrant0607, this was for end users per a convo this morning with @thockin

File: docs/deprecation_policy.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+legacy API groups into more domain-specific API groups.
+
+Rule #2: API objects must be able to round-trip between API versions in a given
+release without information loss, with the exception of whole REST resources
+that do not exist in some versions.
+
+For example, an object can be written as v1 and then read back as v2 and
+converted to v1, and the resulting v1 resource will be identical to the
+original.  The representation in v2 might be different from v1, but the system
+knows how to convert between them in both directions.
```
Comment (mtaufen) [inline docs/deprecation_policy.md]: So if I read this right it is impossible to remove a field that existed in v1 when you move to v2, unless you somehow couple that information somewhere else in the v2 object.

**Comment by bgrant0607** [comment]
@thockin Nit: Please don't use underscores in doc file names. When we move this file to its proper location it should also be renamed.

File: docs/deprecation_policy.md (hunk 3 above)
Comment (bgrant0607) [inline docs/deprecation_policy.md]: So far we've been putting such notices in the release notes.

File: docs/deprecation_policy.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+be round-tripped.  For example a v1 field named “magnitude” which was
+deprecated might be named “deprecatedMagnitude” in API v2.  When v1 is
+eventually removed, the deprecated field can be removed from v2.
+
+### Enumerated or constant values
+
+As with whole REST resources and fields thereof, a constant value which was
+supported in API v1 must exist and function until API v1 is removed.
+
+### Future work
```
Comment (bgrant0607) [inline docs/deprecation_policy.md]: If we think these rules should apply identically to versioned configuration, we should add a subsection that explicitly states that. I think that's a reasonable starting point. I think the only config that's GA might be kubeconfig and the scheduler config.

File: docs/deprecation_policy.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+| X       | v1           |       |
+| X+1     | v1, v2alpha1 |       |
+| X+2     | v1, v2alpha2 | * v2alpha1 is removed, “action required” relnote |
+| X+3     | v1, v2beta1  | * v2alpha2 is removed, “action required” relnote |
+| X+4     | v1, v2beta1, v2beta2 | * v2beta1 is deprecated, “action required” relnote |
+| X+5     | v1, v2, v2beta2      | * v2beta1 is removed, “action required” relnote<br> * v2beta2 is deprecated, “action required” relnote<br> * v1 is deprecated, “action required” relnote |
+| X+6     | v1, v2               | * v2beta2 is removed, “action required” relnote |
+| X+7     | v1, v2               | |
+| X+8     | v1, v2               | |
+| X+9     | v2                   | * v1 is removed, “action required” relnote |
```
Comment (aronchick) [inline docs/deprecation_policy.md]: I thought we had a "2 version" deprecation policy exclusively (no time)? This is much larger than we currently test against.

**Comment by lavalamp** [comment]
You are thinking of binary versions, this table is about API versions.

File: docs/deprecation_policy.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+converted to v1, and the resulting v1 resource will be identical to the
+original.  The representation in v2 might be different from v1, but the system
+knows how to convert between them in both directions.
+
+Rule #3: Other than the “current” API version, older API versions must be
+supported after their announced deprecation for a duration of no less than:
+   * GA: 1 year or 2 releases (whichever is longer)
+   * Beta: 3 months or 1 release (whichever is longer)
+   * Alpha: 0 releases
+
```
Comment (vishh) [inline docs/deprecation_policy.md]: Should we also require releasing additional tooling that will help translate old API objects to newer versions?

File: docs/deprecation_policy.md (hunk 2 above)
Comment (vishh) [inline docs/deprecation_policy.md]: Do we really to support all features or is this a side effect of the round tripping requirement?

File: docs/deprecation_policy.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+We intend to introduce more fine-grained API versions, at which point we will
+adjust these rules as needed.
+
+## Deprecating a flag or CLI
+
+The Kubernetes system is comprised of several different programs cooperating.
+Sometimes, we need to remove flags or CLI commands (collectively “CLI
+elements”) in these programs.  The individual programs naturally sort into two
+main groups - user-facing and admin-facing programs, which vary slightly in
+their deprecation policies.  Unless a flag is explicitly prefixed “alpha” or
```
Comment (vishh) [inline docs/deprecation_policy.md]: How do we handle alpha features that are exposed as top level commands?

File: docs/deprecation_policy.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+## Deprecating a feature or behavior
+
+Occasionally we need to deprecate some feature or behavior of the system that
+is not controlled by the API or CLI.  In this case, the rules for deprecation
+are as follows:
+
+Rule #6: Deprecated behaviors must function for no less than 1 year after their
+announced deprecation.
+
+This does not imply that all changes to the system are governed by this policy.  This applies only to significant, user-visible behaviors which impact the correctness of applications running on Kubernetes or that impact the administration of Kubernetes clusters, and which are being removed entirely.
```
Comment (vishh) [inline docs/deprecation_policy.md]: This subjective policy will lead to confusion. Would it make sense to define a process for deciding deprecation period for such features?

File: docs/deprecation_policy.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+A given release of Kubernetes can support any number of API groups and any
+number of versions of each.
+
+The following rules govern the deprecation of elements of the API.  This
+includes:
+   * REST resources (aka API objects)
+   * Fields of REST resources
+   * Enumerated or constant values
+These rules are enforced between official releases, not between
```
Comment (roberthbailey) [inline docs/deprecation_policy.md]: Need a blank line before this so that it's a new paragraph.

File: docs/deprecation_policy.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+arbitrary commits to master or release branches.
+
+Rule #1: API elements may only be removed by incrementing the version of the
```
Comment (roberthbailey) [inline docs/deprecation_policy.md]: Can you bold or ### each rule so that they stand out a bit more?

File: docs/deprecation_policy.md (hunk 4 above)
Comment (roberthbailey) [inline docs/deprecation_policy.md]: I think none of these rules should apply to alpha apis. They should be able to change in arbitrary ways at any point in time to encourage fast development.

File: docs/deprecation_policy.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+Rule #3: Other than the “current” API version, older API versions must be
+supported after their announced deprecation for a duration of no less than:
+   * GA: 1 year or 2 releases (whichever is longer)
+   * Beta: 3 months or 1 release (whichever is longer)
+   * Alpha: 0 releases
+
+This is best illustrated by example.  Imagine a Kubernetes release, version X,
+which supports a particular API group.  A new Kubernetes release is made every
+3 months (4 per year).  The following table describes which API versions are
```
Comment (roberthbailey) [inline docs/deprecation_policy.md]: nit: made *approximately* every 3 months.

File: docs/deprecation_policy.md [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+supported in a series of subsequent releases.
+
+| Release | API versions | Notes |
+|---------|--------------|-------|
+| X       | v1           |       |
+| X+1     | v1, v2alpha1 |       |
+| X+2     | v1, v2alpha2 | * v2alpha1 is removed, “action required” relnote |
+| X+3     | v1, v2beta1  | * v2alpha2 is removed, “action required” relnote |
+| X+4     | v1, v2beta1, v2beta2 | * v2beta1 is deprecated, “action required” relnote |
+| X+5     | v1, v2, v2beta2      | * v2beta1 is removed, “action required” relnote<br> * v2beta2 is deprecated, “action required” relnote<br> * v1 is deprecated, “action required” relnote |
```
Comment (roberthbailey) [inline docs/deprecation_policy.md]: I don't see it explicitly called out that you can't deprecate v1 until v2 is available (e.g. it shouldn't be possible to deprecate v1 at X+3 because we *almost* have a v2 ready, just to speed up the time at which we can remove v1).

File: docs/deprecation_policy.md (hunk 5 above)
Comment (roberthbailey) [inline docs/deprecation_policy.md]: This is covered below (search for "deprecatedMagnitude").

File: docs/deprecation_policy.md (hunk 1 above)
Comment (thockin) [inline docs/deprecation_policy.md]: I don't understand your comment, @smarterclayton

File: docs/deprecation_policy.md (hunk 3 above)
Comment (thockin) [inline docs/deprecation_policy.md]: release notes seem like a good way to announce them, but I think we need a page that links to current deprecations, no?

File: docs/deprecation_policy.md (hunk 2 above)
Comment (vishh) [inline docs/deprecation_policy.md]: When can features reach EOL?

File: docs/deprecation_policy.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+Rule #2: API objects must be able to round-trip between API versions in a given
+release without information loss, with the exception of whole REST resources
+that do not exist in some versions.
+
+For example, an object can be written as v1 and then read back as v2 and
+converted to v1, and the resulting v1 resource will be identical to the
+original.  The representation in v2 might be different from v1, but the system
+knows how to convert between them in both directions.
+
+Rule #3: Other than the “current” API version, older API versions must be
```
Comment (thockin) [inline docs/deprecation_policy.md]: wilco

File: docs/deprecation_policy.md (hunk 8 above)
Comment (thockin) [inline docs/deprecation_policy.md]: Maybe, but that is covered elsewhere - this doc is about deprecation and nothing more.

File: docs/deprecation_policy.md (hunk 9 above)
Comment (thockin) [inline docs/deprecation_policy.md]: alpha CLI commands is interesting.  They probably should be denoted "alpha" somehow, but I think these rules are sufficient to cover deprecation.

File: docs/deprecation_policy.md (hunk 10 above)
Comment (thockin) [inline docs/deprecation_policy.md]: I could be convinced to add more words here, but I don;t want this to become a legal treatise on the meaning of "significant".

File: docs/deprecation_policy.md (hunk 12 above)
Comment (thockin) [inline docs/deprecation_policy.md]: **ack**

File: docs/deprecation_policy.md (hunk 4 above)
Comment (thockin) [inline docs/deprecation_policy.md]: once we publish an alpha API, I think it's unreasonable to change it in incompatible ways.  Revving APIs should be cheap (I know it's not but @lavalamp is on it :)

File: docs/deprecation_policy.md (hunk 13 above)
Comment (thockin) [inline docs/deprecation_policy.md]: roger roger

**Comment by thockin** [comment]
New push is up with much feedback addressed.  File was renamed, so github may lose comments.

File: docs/deprecation_policy.md (hunk 6 above)
Comment (bgrant0607) [inline docs/deprecation_policy.md]: @thockin I'd do the latter: Call out component configs and say that the same policies apply as those that apply to REST resources.

File: docs/deprecation-policy.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+We intend to introduce more fine-grained API versions, at which point we will
+adjust these rules as needed.
+
+## Deprecating a flag or CLI
+
+The Kubernetes system is comprised of several different programs cooperating.
+Sometimes, we need to remove flags or CLI commands (collectively “CLI
+elements”) in these programs.  The individual programs naturally sort into two
+main groups - user-facing and admin-facing programs, which vary slightly in
+their deprecation policies.  Unless a flag is explicitly prefixed “alpha” or
```
Comment (davidopp) [inline docs/deprecation-policy.md]: Do we have flags that control the behavior of alpha/beta features today? If so what you describe here would require some bulk flag renaming.

File: docs/deprecation-policy.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+“beta”, it is considered GA.
+
+CLI elements are effectively part of the API to the system, but since they are
```
Comment (davidopp) [inline docs/deprecation-policy.md]: I couldn't quite figure out what a "CLI element" is. For kubectl, do you mean a subcommand (like get, drain, proxy, etc.)? What does it mean for kubelet?

**Comment by thockin** [comment]
ACK

**Comment by thockin** [comment]
yeah, we have not been 100% consistent on naming.  Will add "or documented"

**Comment by thockin** [comment]
I defined it in the previous paragraph

**Comment by bgrant0607** [comment]
@thockin 

LGTM. I'm favor of merging whenever you feel like you've addressed enough of the comments and iterating more in subsequent PRs.

Any idea why netlify failed?

**Comment by thockin** [comment]
I'm waiting for Devin to do a docs pass and to tell me where he wants the
doc to live and where to link to it from :)

**Comment by davidopp** [comment]
> I defined it in the previous paragraph

I think part of my confusion is whether "e.g. kubectl" is modifying "user-facing components" or "CLI elements" (and likewise whether "e.g. kubelet" is modifying "admin-facing components" or "CLI elements."

TBH I stll don't know what a CLI element is even after re-reading. You say "CLI elements are effectively part of the API" but that doesn't tell me what they are.

**Comment by thockin** [comment]
```
The Kubernetes system is comprised of several different programs
cooperating.
Sometimes, we need to remove flags or CLI commands (collectively “CLI
elements”) in these programs.
```

is that not clear?  Suggest rewording?

On Tue, Dec 6, 2016 at 12:23 PM, David Oppenheimer <notifications@github.com

File: docs/deprecation-policy.md [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+exist and function in releases up to and including X+8.  Only in release X+9,
+when API v1 has aged out, does the Widget resource cease to exist, and the
+behavior get removed.
+
+### Fields of REST resources
+
+As with whole REST resources, an individual field which was present in API v1
+must exist and function until API v1 is removed.  Unlike whole resources, the
+v2 APIs may choose a different representation for the field, as long as it can
+be round-tripped.  For example a v1 field named “magnitude” which was
```
Comment (soltysh) [inline docs/deprecation-policy.md]: I think a different example would be better, eg. v1 has `magnitude`, which transforms into `magnitude1` and `magnitude1` in v2. The round-trip knows how to deal with that, obviously. That's what will usually happen, imho. Adding `deprecatedField` in v2 to be then removed when v1 is gone, is confusing.

File: docs/deprecation-policy.md [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+
+Once an API element has been added to an API group at a particular version, it
+can not be removed from that version or have its behavior significantly
+changed, regardless of track.
+
+Note: For historical reasons, we have 2 “monolithic” API groups - “core” (no
+group name) and “extensions”.  We will incrementally move resources from these
+legacy API groups into more domain-specific API groups.
+
+**Rule #2: API objects must be able to round-trip between API versions in a given
```
Comment (derekwaynecarr) [inline docs/deprecation-policy.md]: @kubernetes/sig-autoscaling -- this was the point I was trying to find in our meeting today.  @DirectXMan12 my reading of this is that HPA v2 would need to convert to HPA v1 when handling scaling on metrics other than cpu.

Comment (DirectXMan12) [inline docs/deprecation-policy.md]: If that's the case, this should probably provide an explicit "if it adds new features, you have to have corresponding annotations on v1" or something like that.

Comment (DirectXMan12) [inline docs/deprecation-policy.md]: (as in, an official policy for what to do when v1 is insufficiently expressive)

Comment (derekwaynecarr) [inline docs/deprecation-policy.md]: see annotation example here: https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions

Comment (fgrzadkowski) [inline docs/deprecation-policy.md]: @thockin What's the policy for adding completely new features in a new API version? Example we have in mind is adding support for custom metrics in v2 HPA API.

**Comment by thockin** [comment]
@soltysh, but that's what I am trying to illustrate - how do you get rid of
a field?

**Comment by thockin** [comment]
@DirectXMan12 fair point.

**Comment by thockin** [comment]
@fgrzadkowski The current rule is that it must be able to round-trip.
Consider if you add something in v2, the user uses that thing, but your
cluster is still using v1 as a storage-version.

**Comment by soltysh** [comment]
> @soltysh, but that's what I am trying to illustrate - how do you get rid of a field?

Right, nvmd, I misread it. Sorry. This LGTM.

**Comment by thockin** [comment]
Now that 1.5 is out, maybe @devin-donnelly can pour some brain power into the finer points of where this should live and whether humans can read it. :)

File: docs/deprecation-policy.md [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+assignees:
+- bgrant0607
+- lavalamp
+- thockin
+
+---
+
+# Kubernetes Deprecation Policy
+
+Kubernetes is a large system with many components and many contributors.  As
```
Comment (devin-donnelly) [inline docs/deprecation-policy.md]: Try to avoid "we" constructs.

"Kubernetes is a large system with many components and contributors. As with any such software, the feature set evolves over time, and sometimes a feature may need to be removed. This could include an API, a flag, or even an entire feature. To avoid making breaking changes for existing users, Kubernetes is instituting a *deprecation policy* for aspects of the system that are slated to be removed."

File: docs/deprecation-policy.md [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+with any such software, the feature set naturally evolves over time.  Sometimes
+we realize that something we have done in the past, whether that is some API, a
+flag, or even a whole feature, needs to be replaced or removed.  Kubernetes is
+committed to not breaking users, so we can’t just make breaking changes.  This
+is when deprecation is triggered.
+
+This document details the policies we hold on deprecation for various facets of
```
Comment (devin-donnelly) [inline docs/deprecation-policy.md]: "This document details the policies for deprecating the different facets of the Kubernetes system should they be removed in future releases."

File: docs/deprecation-policy.md [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+the system.
+
+## Deprecating parts of the API
+
+Since Kubernetes is an API-driven system, it’s not surprising that the API
```
Comment (devin-donnelly) [inline docs/deprecation-policy.md]: "We/our" construct.

"Since Kubernetes is an API-driven system, the API has evolved over time to reflect an evolving understanding of the problem space. The Kubernetes API consists of several "API groups", and each API group is independently versioned."

File: docs/deprecation-policy.md [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+|----------|----------------------------------|
+| v1       | GA (generally available, stable) |
+| v1beta1  | Beta (pre-release)               |
+| v1alpha1 | Alpha (experimental)             |
+
+A given release of Kubernetes can support any number of API groups and any
+number of versions of each.
+
+The following rules govern the deprecation of elements of the API.  This
+includes:
```
Comment (devin-donnelly) [inline docs/deprecation-policy.md]: Needs a blank line underneath to make the bullet list render correctly.

File: docs/deprecation-policy.md [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+Component configs are versioned and managed just like REST resources.
+
+### Future work
+
+We intend to introduce more fine-grained API versions, at which point we will
+adjust these rules as needed.
+
+## Deprecating a flag or CLI
+
+The Kubernetes system is comprised of several different programs cooperating.
```
Comment (devin-donnelly) [inline docs/deprecation-policy.md]: "we" construct.

"Sometimes, a Kubernetes release might remove flags or CLI commands"

**Comment by devin-donnelly** [comment]
@thockin A few wording nits. Otherwise, no problems.

**Comment by thockin** [comment]
All "we" have been hunted down and exterminated.  Is this the right dir for
this file?

This proposes http://kubernetes.io/docs/deprecated/ as a dir for the
various deprecation notices.  OK?

**Comment by devin-donnelly** [comment]
@thockin Yes, this file is fine in the root directory, and I'm fine with the location for Deprecation notices. Anybody else need to sign off on this before I can merge it?

**Comment by thockin** [comment]
Nope!  Thanks.  I'll start directing deprecation notice PRs to
http://kubernetes.io/docs/deprecated/

**Comment by thockin** [comment]
rebased and repushed.  I do not have access to the netlify, but it said it failed before.

**Comment by thockin** [comment]
rebased and pushed again.  Gonna self-merge
